File: src/xianyu_cli/models/envelope.py
"""Structured output envelope for all CLI commands.

All commands return a consistent JSON envelope:
  {"ok": bool, "schema_version": str, "data": ..., "error": ...}

Rich output goes to stderr, JSON to stdout — pipe-safe by design.
"""

from __future__ import annotations

import json
import sys
from dataclasses import asdict, dataclass, field
from typing import Any

from rich.panel import Panel
from rich.text import Text

from xianyu_cli.utils._common import console

SCHEMA_VERSION = "1.0.0"


@dataclass
class Envelope:
    ok: bool
    data: Any = None
    error: str | None = None
    schema_version: str = field(default=SCHEMA_VERSION)

    def to_dict(self) -> dict:
        return asdict(self)

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=2)

    def emit(self, output_mode: str = "rich") -> None:
        """Emit the envelope. JSON → stdout, Rich → stderr."""
        if output_mode == "json":
            sys.stdout.write(self.to_json() + "\n")
            sys.stdout.flush()
        else:
            if self.ok:
                self._emit_rich_success()
            else:
                self._emit_rich_error()

    def _emit_rich_success(self) -> None:
        if self.data is None:
            console.print("[green]OK[/green]")
            return
        if isinstance(self.data, str):
            console.print(self.data)
        elif isinstance(self.data, dict):
            from rich.pretty import Pretty
            console.print(Pretty(self.data))
        elif isinstance(self.data, list):
            from rich.pretty import Pretty
            console.print(Pretty(self.data))
        else:
            console.print(str(self.data))

    def _emit_rich_error(self) -> None:
        msg = Text(f"Error: {self.error}", style="bold red")
        console.print(Panel(msg, title="[red]Failed[/red]", border_style="red"))


def ok(data: Any = None) -> Envelope:
    """Create a success envelope."""
    return Envelope(ok=True, data=data)


def fail(error: str) -> Envelope:
    """Create an error envelope."""
    return Envelope(ok=False, error=error)

File: golang/models/envelope.go
package models

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/fatih/color"
)

const SchemaVersion = "1.0.0"

// Envelope is the unified output wrapper for all CLI commands.
type Envelope struct {
	OK            bool        `json:"ok"`
	Data          interface{} `json:"data,omitempty"`
	Error         string      `json:"error,omitempty"`
	SchemaVersion string      `json:"schema_version"`
}

// OK creates a success envelope.
func OK(data interface{}) *Envelope {
	return &Envelope{OK: true, Data: data, SchemaVersion: SchemaVersion}
}

// Fail creates an error envelope.
func Fail(errMsg string) *Envelope {
	return &Envelope{OK: false, Error: errMsg, SchemaVersion: SchemaVersion}
}

// ToJSON serializes the envelope to JSON.
func (e *Envelope) ToJSON() string {
	data, err := json.MarshalIndent(e, "", "  ")
	if err != nil {
		return fmt.Sprintf(`{"ok":false,"error":"json marshal error: %s","schema_version":"%s"}`, err.Error(), SchemaVersion)
	}
	return string(data)
}

// Emit outputs the envelope in the specified format.
// JSON goes to stdout, Rich goes to stderr.
func (e *Envelope) Emit(outputMode string) {
	if outputMode == "json" {
		fmt.Fprintln(os.Stdout, e.ToJSON())
	} else {
		if e.OK {
			e.emitRichSuccess()
		} else {
			e.emitRichError()
		}
	}
}

func (e *Envelope) emitRichSuccess() {
	if e.Data == nil {
		fmt.Fprintln(os.Stderr, color.GreenString("OK"))
		return
	}
	switch v := e.Data.(type) {
	case string:
		fmt.Fprintln(os.Stderr, v)
	case map[string]interface{}:
		if msg, ok := v["message"]; ok {
			fmt.Fprintln(os.Stderr, color.GreenString("✓")+" "+fmt.Sprint(msg))
		} else {
			data, _ := json.MarshalIndent(v, "", "  ")
			fmt.Fprintln(os.Stderr, string(data))
		}
	default:
		data, err := json.MarshalIndent(v, "", "  ")
		if err != nil {
			fmt.Fprintln(os.Stderr, fmt.Sprint(v))
		} else {
			fmt.Fprintln(os.Stderr, string(data))
		}
	}
}

func (e *Envelope) emitRichError() {
	red := color.New(color.FgRed, color.Bold)
	fmt.Fprintf(os.Stderr, "\n%s %s\n\n", red.Sprint("Error:"), e.Error)
}
